Created SequenceLexicoder to resolve #1252

reggert opened the pull request:
This provides a replacement for `ListLexicoder` that supports empty lists, resolving #1252 .

The branch is based on the `master` branch, but I can rebase it to a different branch if necessary to include it in the 2.0 release.

I'm also open to suggestions for a different name for the class if anyone can think of a better one.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/client/lexicoder/SequenceLexicoderTest.java:
```diff
@@ -0,0 +1,98 @@
+  private final List<Long> data5 = new ArrayList<>();
+
+  @Before
+  public void setUp() {
```
Would be nice if the data had empty elements in order to test the sort order of list with empty elements vs empty list.

reggert replied:
I just copied `ListLexicoder` test and added an empty list to the data sets. I can add a case for empty elements, but it will require changing the element type from `Long` to something with variable since like `String`.

keith-turner replied:
Thanks for doing this in the new PR

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/lexicoder/SequenceLexicoder.java:
```diff
@@ -0,0 +1,78 @@
+      byte[] escapedElement = escapedElements[i];
+      ret.add(lexicoder.decode(unescape(escapedElement)));
+    }
+
```
Could throw an exception if the last ignored element in `escapedElements` is not empty byte array.

reggert replied:
True. It might make a useful sanity check to catch cases where the data is not formatted as expected, though we otherwise wouldn't really care about the last element.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/lexicoder/SequenceLexicoder.java:
```diff
@@ -0,0 +1,78 @@
+  protected List<LT> decodeUnchecked(byte[] b, int offset, int len) {
+
+    byte[][] escapedElements = split(b, offset, len);
+    assert escapedElements.length > 0;
```
Could throw an exception here.

reggert replied:
I thought about this, but then realized that `split` will always return at least one element, even if the input is empty. I added the assertion to validate this invariant. I'll add a comment to it to explain the reasoning.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/lexicoder/ListLexicoder.java:
```diff
@@ -30,6 +30,8 @@
  * encoded element sorts lexicographically.
  *
+ * Note: Empty lists are not supported.
+ *
```
This could use the `@see` javadoc tag to point to SequenceLexicoder

Also, could make the encode method throw an exception when encoding an empty list.  The exception could have a message saying not support and mention the new encoded.

reggert replied:
Agreed.

reggert commented:
I've addressed the issues and cleaned up the code a bit. I've also re-based onto the 2.0 branch to avoid being dependent upon Java 11. I'll open a new PR with the updated merge target.
